server/src/main/java/io/crate/expression/symbol/Symbol.java

```diff
@@ -197,14 +197,19 @@ default Symbol cast(DataType<?> targetType, CastMode... modes) {
             return this;
         } else if (targetType.equals(valueType())) {
             return this;
-        } else if (ArrayType.unnest(targetType).equals(DataTypes.UNTYPED_OBJECT)
-                   && valueType().id() == targetType.id()) {
-            return this;
-        } else if (ArrayType.unnest(targetType).equals(DataTypes.NUMERIC)
-                   && valueType().id() == DataTypes.NUMERIC.id()) {
-            // Do not cast numerics to unscaled numerics because we do not want to loose precision + scale
-            return this;
+        } else {
+            DataType<?> innerTargetType = ArrayType.unnest(targetType);
+            if (innerTargetType.equals(DataTypes.UNTYPED_OBJECT) &&
+                innerTargetType.id() == ArrayType.unnest(valueType()).id() &&
```

Comment:
I'd prefer this for readability
```suggestion
            DataType<?> innerTargetType = ArrayType.unnest(targetType);
            DataType<?> innerValueType =  ArrayType.unnest(valueType());
            if (innerTargetType.equals(DataTypes.UNTYPED_OBJECT) &&
                innerTargetType.id() == innerValueType.id() &&
```

Reply:
We can do that, but if the `innerTargetType` is not an `UNTYPED_OBJECT` there is no need to unest the `valueType`, therefore this expression is more efficient.